library/psa_crypto.c

```diff
@@ -713,118 +713,187 @@ static psa_status_t psa_import_rsa_key( psa_key_slot_t *slot,
          * So its data length is 2m+1 where n is the key size in bits.
          */
-        if( ( data_length & 1 ) == 0 )
+        if( ( slot->data.key.bytes & 1 ) == 0 )
             return( PSA_ERROR_INVALID_ARGUMENT );
-        data_length = data_length / 2;
+        data_length = slot->data.key.bytes / 2;
     }
 
     /* Load the group. */
-    grp_id = mbedtls_ecc_group_of_psa( curve, data_length );
+    grp_id = mbedtls_ecc_group_of_psa( PSA_KEY_TYPE_ECC_GET_FAMILY( slot->attr.type),
+                                       data_length );
     if( grp_id == MBEDTLS_ECP_DP_NONE )
         return( PSA_ERROR_INVALID_ARGUMENT );
-    return( mbedtls_to_psa_error(
-                mbedtls_ecp_group_load( &( *p_ecp )->grp, grp_id ) ) );
-}
+    status = mbedtls_to_psa_error(
+                mbedtls_ecp_group_load( &ecp->grp, grp_id ) );
+    if( status != PSA_SUCCESS )
+        return( status );
 
-/* Import a public key given as the uncompressed representation defined by SEC1
- * 2.3.3 as the content of an ECPoint. */
-static psa_status_t psa_import_ec_public_key( psa_ecc_family_t curve,
-                                              const uint8_t *data,
-                                              size_t data_length,
-                                              mbedtls_ecp_keypair **p_ecp )
-{
-    psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
-    mbedtls_ecp_keypair *ecp = NULL;
+    /* Load the key material */
+    if( PSA_KEY_TYPE_IS_PUBLIC_KEY( slot->attr.type ) )
+    {
+        /* Load the public value. */
+        status = mbedtls_to_psa_error(
+            mbedtls_ecp_point_read_binary( &ecp->grp, &ecp->Q,
+                                           slot->data.key.data,
+                                           slot->data.key.bytes ) );
+        if( status != PSA_SUCCESS )
+            goto exit;
 
-    status = psa_prepare_import_ec_key( curve, data_length, 1, &ecp );
-    if( status != PSA_SUCCESS )
-        goto exit;
+        /* Check that the point is on the curve. */
+        status = mbedtls_to_psa_error(
+            mbedtls_ecp_check_pubkey( &ecp->grp, &ecp->Q ) );
+        if( status != PSA_SUCCESS )
+            goto exit;
+    }
+    else
+    {
+        /* Load the secret value. */
+        status = mbedtls_to_psa_error(
+            mbedtls_ecp_read_key( ecp->grp.id,
+                                  ecp,
+                                  slot->data.key.data,
+                                  slot->data.key.bytes ) );
 
-    /* Load the public value. */
-    status = mbedtls_to_psa_error(
-        mbedtls_ecp_point_read_binary( &ecp->grp, &ecp->Q,
-                                       data, data_length ) );
+        if( status != PSA_SUCCESS )
+            goto exit;
+        /* Validate the private key. */
+        status = mbedtls_to_psa_error(
+            mbedtls_ecp_check_privkey( &ecp->grp, &ecp->d ) );
+        if( status != PSA_SUCCESS )
+            goto exit;
+    }
```

Comment:
In my mind, this is performance-enhancing behaviour: if you're loading a private key and only need the private part, why would we spend cycles on calculating the public part? I'd argue that the 99% use case for ECP keys is verifying with a public key, or signing with a private key. That means the need for calculating the public part from the private key is in reality only when generating an ephemeral/identity key and exporting the public part to give to a counterpart. Since calculating the public key from a private key is a non-trivial operation, doing this only when absolutely required is IMO best practice.

IIRC there was also a comment somewhere (forgot whether it was here or on the mailing list) where someone using mbedTLS on an MCU-type target was complaining about always having to recalculate the public key when it isn't necessary.

I don't agree that this needs to be a separate commit. The previous behaviour was IMO guided by having the internal representation inside the key slot, where you were expecting to just have to pass the representation from the keyslot into the mbedtls_ functions. With this change, there is no such expectation anymore. You're now just-in-time loading the representation of exactly what was imported into the slot, which is either a public key or a private key. That information is inside the slot (and in the representation...), so there's really no excuse to *not* do this in conjunction with the slot content redefinition.

Reply:
I think Mbed TLS requires the public key during private key operations, so if no accelerator is used, the public key has to be computed anyway before doing anything with the key except export-private. Where is this done now?

Avoiding computing the public key when not necessary (which is only when using an accelerator) is an optimization that doesn't belong in this rework.

Reply:
The two first paragraphs provide to me good and useful information about the rationale of the change. And it would be beneficial to have that information in the project history thus I do think that it would better for this to be in a separate commit. Given what you say in the 3rd paragraph, it would be in a commit after this commit (not before as I suggested before) though. With a dedicated commit you can put the explanations above in the commit message. From paragraph two, probably that a entry in the change log for users would probably be useful. This change log entry would fit nicely in the dedicated commit :-).

Reply:
> I think Mbed TLS requires the public key during private key operations, so if no accelerator is used, the public key has to be computed anyway before doing anything with the key except export-private. Where is this done now?
I can't find any documentation that says it does. And the actual implementation definitely doesn't (I just checked). ecp->Q is only accessed (read from) in these functions:
* export pubkey (pk write)
* ECDSA verify (ecdsa)
* `check_pub_priv` (ECP)
* ECDH `get_params_internal` copies key->Q to ctx->Q (which doesn't fail with a zero Q), but ctx->Q is not accessed later on in the ECDH process. In contrast, it explicitly provides `make_public` to calculate Q from d.

To me, the fact that there is no requirement written down that Q must be present for operations which only require d, plus the fact that ECDH was apparently written with this already covered, plus the actual behaviour of the library (only accessing Q on operations specifically requiring a public key) points to there not being such a requirement, neither official nor implicit.

> Avoiding computing the public key when not necessary (which is only when using an accelerator) is an optimization that doesn't belong in this rework.
Disagree, given that mbedTLS doesn't require it, it benefits all and not just accelerators.

Reply:
If we are sure of the points where the public key has to be computed (two for the time being) I am ok with this change. If we keep it, this probably deserves a change log.

Reply:
Ok, Mbed TLS doesn't actually require the public key. All the official ways to build an `mbedtls_ecp_keypair` fill in the public key, but the code that uses private keys doesn't actually require it, and `mbedtls_ecp_keypair` is defined in a public header so the necessary invariants are not clear. Given that this is only for an object that isn't exposed externally anyway, I'm ok with not filling in the public key here. If this became a concrete problem, functional testing would reveal it easily, so it doesn't add much maintenance overhead.